### PR TITLE
Path build and onioned messages

### PR DESCRIPTION
**Pull request by tewinget**
WIP

### CODE REVIEW

File: llarp/path/path.cpp [hunk 1]
```diff
@@ -107,8 +107,33 @@ namespace llarp::path
...
+    btdp.append("BODY", body);
+    auto payload = std::move(btdp).str();
+    auto* payload_ptr = reinterpret_cast<unsigned char*>(payload.data());
+
+    // TODO: old impl padded messages if smaller than a certain size; do we still want to?
+
+    auto crypto = CryptoManager::instance();
+
+    TunnelNonce nonce;
+    outer_nonce.Randomize();
```
Comment (dr7ana) [inline llarp/path/path.cpp:121]: Where does this come from?

File: llarp/path/path.cpp [hunk 2]
```diff
@@ -107,8 +107,33 @@ namespace llarp::path
   Path::send_path_control_message(
       std::string method, std::string body, std::function<void(oxen::quic::message m)> func)
   {
+    oxenc::bt_dict_producer btdp;
+    btdp.append("METHOD", method);
+    btdp.append("BODY", body);
+    auto payload = std::move(btdp).str();
```
Comment (dr7ana) [inline llarp/path/path.cpp:113]: Might be good to wrap the BTDP in braces before you std move it out of scope like you did in the other logic

File: llarp/path/path.cpp [hunk 3]
```diff
@@ -107,8 +107,33 @@ namespace llarp::path
...
+    outer_nonce.Randomize();
+    for (const auto& hop : hops)
+    {
+      // do a round of chacha for each hop and mutate the nonce with that hop's nonce
+      CryptoManager::instance()->xchacha20(payload_ptr, hop.shared, nonce);
+      nonce ^= hop.nonceXOR;
+    }
+
+    oxenc::bt_dict_producer outer_dict;
+    outer_dict.append("PATHID", TXID().ToView());
```
Comment (dr7ana) [inline llarp/path/path.cpp:130]: Small note: would be best to alphabetize these keys (same with the ones at the top)

File: llarp/path/pathbuilder.cpp [hunk 4]
```diff
@@ -465,17 +465,44 @@ namespace llarp
...
+      // Path build fail and success are handled poorly at best and changing how we
+      // handle these responses as well as how we store and use Paths as a whole might
+      // be worth doing sooner rather than later.  Leaving some TODOs below where fail
+      // and success live.
+      auto response_cb = [self](oxen::quic::message m) {
+        try
+        {
+          if (not m)
+          {
+            if (m.timed_out)
```
Comment (dr7ana) [inline llarp/path/pathbuilder.cpp:478]: This is awkward

File: llarp/link/link_manager.hpp [hunk 5]
```diff
@@ -286,23 +286,21 @@ namespace llarp
...
    private:
     // DHT messages
-    void handle_find_name(oxen::quic::message);      // relay
-    void handle_find_intro(oxen::quic::message);     // relay
-    void handle_publish_intro(oxen::quic::message);  // relay
-    void handle_find_router(oxen::quic::message);    // relay + path
+    void handle_find_name(oxen::quic::message, const RouterID& from);      // relay
+    void handle_find_intro(oxen::quic::message, const RouterID& from);     // relay
+    void handle_publish_intro(oxen::quic::message, const RouterID& from);  // relay
+    void handle_find_router(oxen::quic::message, const RouterID& from);    // relay + path
```
Comment (dr7ana) [inline llarp/link/link_manager.hpp:292]: Why did you add this to all of them?

File: llarp/path/path.cpp [hunk 6]
```diff
@@ -107,8 +107,33 @@ namespace llarp::path
   Path::send_path_control_message(
       std::string method, std::string body, std::function<void(oxen::quic::message m)> func)
   {
+    oxenc::bt_dict_producer btdp;
+    btdp.append("METHOD", method);
+    btdp.append("BODY", body);
+    auto payload = std::move(btdp).str();
+    auto* payload_ptr = reinterpret_cast<unsigned char*>(payload.data());
```
Comment (dr7ana) [inline llarp/path/path.cpp:114]: There's no benefit to doing this outside of the function call as its only used once

File: llarp/path/path.cpp [hunk 7]
```diff
@@ -107,8 +107,33 @@ namespace llarp::path
...
+    // TODO: old impl padded messages if smaller than a certain size; do we still want to?
+
+    auto crypto = CryptoManager::instance();
+
+    TunnelNonce nonce;
+    outer_nonce.Randomize();
+    for (const auto& hop : hops)
+    {
+      // do a round of chacha for each hop and mutate the nonce with that hop's nonce
+      CryptoManager::instance()->xchacha20(payload_ptr, hop.shared, nonce);
```
Comment (dr7ana) [inline llarp/path/path.cpp:125]: This take 4 parameters

File: llarp/path/path.cpp [hunk 8]
```diff
@@ -107,8 +107,33 @@ namespace llarp::path
...
+
+    oxenc::bt_dict_producer outer_dict;
+    outer_dict.append("PATHID", TXID().ToView());
+    outer_dict.append("NONCE", nonce.ToView());
+    outer_dict.append("PAYLOAD", payload);
+
+
     return router.send_control_message(
-        upstream(), std::move(method), std::move(body), std::move(func));
+        upstream(), "path_control", std::move(outer_dict.str()), std::move(func));
```
Comment (dr7ana) [inline llarp/path/path.cpp:136]: `std::move(outer_dict).str()`

File: llarp/path/path.cpp [hunk 9]
```diff
@@ -107,8 +107,33 @@ namespace llarp::path
...
   {
+    oxenc::bt_dict_producer btdp;
+    btdp.append("METHOD", method);
+    btdp.append("BODY", body);
+    auto payload = std::move(btdp).str();
+    auto* payload_ptr = reinterpret_cast<unsigned char*>(payload.data());
+
+    // TODO: old impl padded messages if smaller than a certain size; do we still want to?
+
+    auto crypto = CryptoManager::instance();
```
Comment (dr7ana) [inline llarp/path/path.cpp:118]: Unused

File: llarp/path/path.cpp (hunk 8 above)
Comment (dr7ana) [inline llarp/path/path.cpp:136]: Also, if you're going to create a new endpoint "path_control", whatever `std::function` is passed as `func` needs to be captured by a lambda in the call to `router.send_control_message(...)`

That way any path-specific logic doesn't have to be implemented every single place a function that calls `send_path_control_message` is called, like `find_router` or `find_name`

File: llarp/path/path.cpp [hunk 10]
```diff
@@ -107,8 +107,33 @@ namespace llarp::path
   Path::send_path_control_message(
       std::string method, std::string body, std::function<void(oxen::quic::message m)> func)
   {
+    oxenc::bt_dict_producer btdp;
+    btdp.append("METHOD", method);
+    btdp.append("BODY", body);
```
Comment (jagerman) [inline llarp/path/path.cpp:112]: Alphabetical order!

File: llarp/path/path.cpp (hunk 3 above)
Comment (jagerman) [reply llarp/path/path.cpp:130]: Not just best, it's required.  (It will fail an assert in a debug build)

File: llarp/path/path.cpp (hunk 1 above)
Comment (tewinget) [reply llarp/path/path.cpp:121]: typo, needed 2 different nonces in path build message construction and messed up copying that bit over.

**Comment by tewinget** [comment]
supplanted by #2216
